examples/tree-view/src/Node/treeSlice.js

```diff
@@ -0,0 +1,67 @@
+import { createSlice } from "@reduxjs/toolkit";
+import generateTree from "../generateTree";
+
+const { tree, idCount } = generateTree();
+
+let nextId = idCount;
+const initialState = tree;
+
+const getAllDescendants = (tree, id) => {
+  const descendants = [id];
+  const run = (id) => {
+    tree[id].childIds.forEach((childId) => {
+      descendants.push(childId);
+      run(childId);
+    });
+  };
+
+  run(id);
+  return descendants;
+};
+
+const deleteMany = (tree, ids) => {
+  ids.forEach((id) => delete tree[id]);
+};
+
+const treeSlice = createSlice({
+  name: "tree",
+  initialState,
+  reducers: {
+    increment(state, action) {
+      const nodeId = action.payload;
+      state[nodeId].counter++;
+    },
+    createNode(state) {
+      nextId++;
+      state[nextId] = { id: nextId, counter: 0, childIds: [] };
+    },
+    deleteNode(state, action) {
+      const nodeId = action.payload;
+      const ids = getAllDescendants(state, nodeId);
```

Comment:
:question: Is there a specific need to move this logic outside the reducer? It's not wrong to do so, and naming things can help provide clarity, but it also can make it harder to follow what's going on.